library/Vanilla/TokenSigningTrait.php

```diff
@@ -0,0 +1,199 @@
+<?php
+/**
+ * @author Chris Chabilall chris.c@vanillaforums.com
+ * @copyright 2009-2018 Vanilla Forums Inc.
+ * @license http://www.opensource.org/licenses/gpl-2.0.php GNU GPL v2
+ */
+
+namespace Vanilla;
+
+/**
+ * Trait TokenSigningTrait.
+ *
+ * Methods to be used for token generation and signing.
+ */
+trait TokenSigningTrait {
+    /**
+     * @var string $secret The secret used to sign the token.
+     */
+    public $secret;
+    
+    /**
+     * Get the secret.
+     *
+     * @return mixed Returns the secret.
+     */
+    public function getSecret() {
+        return $this->secret;
+    }
+
+    /**
+     * Set the secret.
+     *
+     * @param mixed $secret
+     * @return $this
+     */
+    public function setSecret($secret) {
+        $this->secret = $secret;
+        return $this;
+    }
+
+    /**
+     * Base64 Encode a string, but make it suitable to be passed in a url.
+     *
+     * @param string $str The string to encode.
+     * @return string The encoded string.
+     */
+    protected static function base64urlEncode($str) {
+        return trim(strtr(base64_encode($str), '+/', '-_'), '=');
+    }
+
+    /**
+     * Decode a string that was encoded using base64UrlEncode().
+     *
+     * @param string $str The encoded string.
+     * @return string The decoded string.
+     */
+    protected static function base64urlDecode($str) {
+        return base64_decode(strtr($str, '-_', '+/'));
+    }
+
+    /**
+     * Generate a random token.
+     *
+     * This token will go into the database, but must be signed before being given to the client.
+     *
+     * @return string Returns a 32 character token string.
+     */
+    public function randomToken() {
+        $token = self::base64urlEncode(openssl_random_pseudo_bytes(24));
+        return $token;
+    }
+
+    /**
+     * Verify a token's signature.
+     *
+     * @param string $accessToken The full access token.
+     * @oaran string $tokenIdentifier the model that is calling verify signature.
+     * @param bool $throw Whether or not to throw an exception on a verification error.
+     * @return bool Returns **true** if the token's expiry date and signature is valid or **false** otherwise.
+     * @throws \Exception Throws an exception if the token is invalid and {@link $throw} is **true**.
+     */
+    public function verifyTokenSignature($accessToken, $tokenIdentifier, $throw = false) {
```

Comment:
You should not need the `$tokenIdentifier` parameter. Per the issue:

> Most of the methods can be used as-is, but some of them mention "access token" by name. Turn this into a protected static property or something like that. The trait can use "token" while the models can use "access token" and "nonce".

You would need to have a protected static property on this trait: `$tokenIdentifier` (or whatever). You would then reference that property in these error messages, instead of concatenating in this parameter. Classes implementing this trait could override the property, automatically customizing their error messages.

Reply:
The reason I went with is parameter was because I kept getting .  Maybe i'm not implementing it properly.

Fatal error: VanillaTests\Fixtures\TokenModel and Vanilla\TokenSigningTrait define the same property ($tokenIdentifier) in the composition of VanillaTests\Fixtures\TokenModel. However, the definition differs and is considered incompatible. Class was composed in /srv/vanilla-repositories/vanilla/tests/fixtures/src/TokenModel.php on line 17
unit-test-php-fpm    |

Reply:
i was able to set it in the constructor, nevermind

Reply:
There's a [documented limitation of trait properties](http://php.net/manual/en/language.oop5.traits.php#language.oop5.traits.properties) that I was unaware of:

> If a trait defines a property then a class can not define a property with the same name unless it is compatible (same visibility and initial value), otherwise a fatal error is issued.

Updating the value in the constructor is a valid way to get around this, but if we're going that route, let's not make it a static property. It doesn't make a lot of sense to me for Vanilla to update a static property every time the object is instantiated.